src/Microsoft.ML.Data/Transforms/KeyToValueTransform.cs

```diff
@@ -67,386 +79,547 @@ private static VersionInfo GetVersionInfo()
-            Source.Schema.GetMetadata<VBuffer<TValue>>(MetadataUtils.Kinds.KeyValues, Infos[iinfo].Source, ref keyMetadata);
-            Host.Check(keyMetadata.Length == typeKey.ItemType.KeyCount);
-
-            VBufferUtils.Densify<TValue>(ref keyMetadata);
-            return new KeyToValueMap<TKey, TValue>(this, typeKey.ItemType.AsKey, typeVal.ItemType.AsPrimitive, keyMetadata.Values, iinfo);
-        }
 
-        protected override ColumnType GetColumnTypeCore(int iinfo)
-        {
-            Host.Assert(0 <= iinfo & iinfo < _types.Length);
-            var type = _types[iinfo];
-            Host.Assert(type != null);
-            return type;
-        }
-
-        protected override Delegate GetGetterCore(IChannel ch, IRow input, int iinfo, out Action disposer)
-        {
-            Host.AssertValueOrNull(ch);
-            Host.AssertValue(input);
-            Host.Assert(0 <= iinfo && iinfo < Infos.Length);
-            disposer = null;
-
-            return _kvMaps[iinfo].GetMappingGetter(input);
-        }
+            public override RowMapperColumnInfo[] GetOutputColumns()
+            {
+                var result = new RowMapperColumnInfo[_parent.ColumnPairs.Length];
+                for (int i = 0; i < _parent.ColumnPairs.Length; i++)
+                {
+                    var meta = RowColumnUtils.GetMetadataAsRow(InputSchema, ColMapNewToOld[i],
+                        x => x == MetadataUtils.Kinds.SlotNames);
+                    result[i] = new RowMapperColumnInfo(_parent.ColumnPairs[i].output, _types[i], meta);
+                }
+                return result;
+            }
 
-        /// <summary>
-        /// A map is an object capable of creating the association from an input type, to an output
-        /// type. This mapping is constructed from key metadata, with the input type being the key type
-        /// and the output type being the type specified by the key metadata.
-        /// </summary>
-        private abstract class KeyToValueMap
-        {
-            /// <summary>
-            /// The item type of the output type, that is, either the output type or,
-            /// if a vector, the item type of that type.
-            /// </summary>
-            protected readonly PrimitiveType TypeOutput;
+            protected override Delegate MakeGetter(IRow input, int iinfo, out Action disposer)
+            {
+                Host.AssertValue(input);
+                Host.Assert(0 <= iinfo && iinfo < _types.Length);
+                disposer = null;
+                return _kvMaps[iinfo].GetMappingGetter(input);
+            }
 
-            /// <summary>
-            /// The column index in Infos.
-            /// </summary>
-            protected readonly int InfoIndex;
+            // Computes the types of the columns and constructs the kvMaps.
+            private void ComputeKVMapsAndMetadata(ISchema schema, out ColumnType[] types, out KeyToValueMap[] kvMaps)
```

Comment:
>ComputeKVMapsAndMetadata [](start = 25, length = 24)

What happened with SlotNames which we used to pass in metadata?

Reply:
They still are

---
In reply to: [216509712](https://github.com/dotnet/machinelearning/pull/856#discussion_r216509712) [](ancestors = 216509712)

Reply:
This is the purpose of line 176

---
In reply to: [216509968](https://github.com/dotnet/machinelearning/pull/856#discussion_r216509968) [](ancestors = 216509968,216509712)

Reply:
Oh, it get moved to GetOutputColumns.
Can we in this case rename it to ComputeKVMaps since it doesn't computeMetadata anymore?

---
In reply to: [216509712](https://github.com/dotnet/machinelearning/pull/856#discussion_r216509712) [](ancestors = 216509712)

Reply:
We sure can, and should :)

---
In reply to: [216510062](https://github.com/dotnet/machinelearning/pull/856#discussion_r216510062) [](ancestors = 216510062,216509712)